File: _archived/notifications-nodejs/src/services/usage-evaluation.service.ts
import { TenantRateLimitPolicyRepository } from "../repositories/tenant-rate-limit-policy.repository";
import { UsageMeterEventRepository } from "../repositories/usage-meter-event.repository";
import { auditClient } from "../integrations/audit/audit.client";
import { logger } from "../shared/logger";
import { NotificationChannel, EnforcementDecision } from "../types";

const rateLimitRepo = new TenantRateLimitPolicyRepository();
const usageRepo = new UsageMeterEventRepository();

function startOfDay(date: Date): Date {
  const d = new Date(date);
  d.setUTCHours(0, 0, 0, 0);
  return d;
}

function startOfMonth(date: Date): Date {
  const d = new Date(date);
  d.setUTCDate(1);
  d.setUTCHours(0, 0, 0, 0);
  return d;
}

function oneMinuteAgo(): Date {
  return new Date(Date.now() - 60_000);
}

/**
 * Check whether a new notification request is allowed for a tenant/channel.
 * Returns allowed=true by default when no policy exists (fail-open on missing policy).
 */
export async function checkRequestAllowed(
  tenantId: string,
  channel: NotificationChannel
): Promise<EnforcementDecision> {
  try {
    const policies = await rateLimitRepo.findActivePolicies(tenantId, channel);
    if (policies.length === 0) return { allowed: true };

    const now = new Date();

    for (const policy of policies) {
      // ── Per-minute request cap ──────────────────────────────────────────────
      if (policy.maxRequestsPerMinute !== null) {
        const recentRequests = await usageRepo.countSince(
          tenantId,
          "api_notification_request",
          oneMinuteAgo(),
          policy.channel ?? undefined
        );
        if (recentRequests >= policy.maxRequestsPerMinute) {
          logger.warn("Rate limit exceeded: maxRequestsPerMinute", {
            tenantId,
            channel,
            recentRequests,
            limit: policy.maxRequestsPerMinute,
          });
          await auditClient.publishEvent({
            eventType: "rate_limit.exceeded",
            tenantId,
            channel,
            metadata: { type: "maxRequestsPerMinute", recentRequests, limit: policy.maxRequestsPerMinute },
          });
          return {
            allowed: false,
            reason: `Rate limit exceeded: ${recentRequests}/${policy.maxRequestsPerMinute} requests in the last minute`,
            code: "RATE_LIMIT_EXCEEDED",
          };
        }
      }

      // ── Daily quota ──────────────────────────────────────────────────────────
      if (policy.maxDailyUsage !== null) {
        const todayStart = startOfDay(now);
        const dailyCount = await usageRepo.countSince(
          tenantId,
          "api_notification_request",
          todayStart,
          policy.channel ?? undefined
        );
        if (dailyCount >= policy.maxDailyUsage) {
          logger.warn("Quota exceeded: maxDailyUsage", { tenantId, channel, dailyCount, limit: policy.maxDailyUsage });
          await auditClient.publishEvent({
            eventType: "quota.exceeded",
            tenantId,
            channel,
            metadata: { type: "maxDailyUsage", dailyCount, limit: policy.maxDailyUsage },
          });
          return {
            allowed: false,
            reason: `Daily quota exceeded: ${dailyCount}/${policy.maxDailyUsage} requests today`,
            code: "DAILY_QUOTA_EXCEEDED",
          };
        }
      }

      // ── Monthly quota ────────────────────────────────────────────────────────
      if (policy.maxMonthlyUsage !== null) {
        const monthStart = startOfMonth(now);
        const monthlyCount = await usageRepo.countSince(
          tenantId,
          "api_notification_request",
          monthStart,
          policy.channel ?? undefined
        );
        if (monthlyCount >= policy.maxMonthlyUsage) {
          logger.warn("Quota exceeded: maxMonthlyUsage", { tenantId, channel, monthlyCount, limit: policy.maxMonthlyUsage });
          await auditClient.publishEvent({
            eventType: "quota.exceeded",
            tenantId,
            channel,
            metadata: { type: "maxMonthlyUsage", monthlyCount, limit: policy.maxMonthlyUsage },
          });
          return {
            allowed: false,
            reason: `Monthly quota exceeded: ${monthlyCount}/${policy.maxMonthlyUsage} requests this month`,
            code: "MONTHLY_QUOTA_EXCEEDED",
          };
        }
      }
    }

    return { allowed: true };
  } catch (err) {
    // Missing or broken policy must NOT crash sends — fail open
    logger.error("UsageEvaluation: enforcement check failed, allowing request", { error: String(err), tenantId, channel });
    return { allowed: true };
  }
}

/**
 * Check per-minute attempt cap (called before creating an outbound attempt).
 * Used when maxAttemptsPerMinute is relevant.
 */
export async function checkAttemptAllowed(
  tenantId: string,
  channel: NotificationChannel
): Promise<EnforcementDecision> {
  try {
    const policies = await rateLimitRepo.findActivePolicies(tenantId, channel);
    if (policies.length === 0) return { allowed: true };

    for (const policy of policies) {
      if (policy.maxAttemptsPerMinute !== null) {
        const attemptUnits = channel === "email"
          ? (["email_attempt"] as const)
          : (["sms_attempt"] as const);

        const recentAttempts = await usageRepo.countSinceMultiple(
          tenantId,
          [...attemptUnits],
          oneMinuteAgo(),
          policy.channel ?? undefined
        );
        if (recentAttempts >= policy.maxAttemptsPerMinute) {
          logger.warn("Rate limit exceeded: maxAttemptsPerMinute", {
            tenantId,
            channel,
            recentAttempts,
            limit: policy.maxAttemptsPerMinute,
          });
          return {
            allowed: false,
            reason: `Attempt rate limit exceeded: ${recentAttempts}/${policy.maxAttemptsPerMinute} attempts in last minute`,
            code: "RATE_LIMIT_EXCEEDED",
          };
        }
      }
    }

    return { allowed: true };
  } catch (err) {
    logger.error("UsageEvaluation: attempt check failed, allowing", { error: String(err), tenantId, channel });
    return { allowed: true };
  }
}

File: apps/services/notifications/Notifications.Infrastructure/Services/UsageEvaluationService.cs
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;

namespace Notifications.Infrastructure.Services;

public class UsageEvaluationService : IUsageEvaluationService
{
    private readonly ITenantRateLimitPolicyRepository _rateLimitRepo;
    private readonly IUsageMeterEventRepository _usageRepo;
    private readonly ILogger<UsageEvaluationService> _logger;

    public UsageEvaluationService(ITenantRateLimitPolicyRepository rateLimitRepo, IUsageMeterEventRepository usageRepo, ILogger<UsageEvaluationService> logger)
    {
        _rateLimitRepo = rateLimitRepo;
        _usageRepo = usageRepo;
        _logger = logger;
    }

    public async Task<EnforcementDecision> CheckRequestAllowedAsync(Guid tenantId, string channel)
    {
        try
        {
            var policies = await _rateLimitRepo.FindActivePoliciesAsync(tenantId, channel);
            if (policies.Count == 0) return new EnforcementDecision { Allowed = true };

            var now = DateTime.UtcNow;
            foreach (var policy in policies)
            {
                if (policy.MaxRequestsPerMinute.HasValue)
                {
                    var recent = await _usageRepo.CountSinceAsync(tenantId, "api_notification_request", now.AddMinutes(-1), policy.Channel);
                    if (recent >= policy.MaxRequestsPerMinute.Value)
                    {
                        _logger.LogWarning("Rate limit exceeded: maxRequestsPerMinute {TenantId} {Channel}", tenantId, channel);
                        return new EnforcementDecision { Allowed = false, Reason = $"Rate limit exceeded: {recent}/{policy.MaxRequestsPerMinute} requests in the last minute", Code = "RATE_LIMIT_EXCEEDED" };
                    }
                }
                if (policy.MaxDailyUsage.HasValue)
                {
                    var todayStart = now.Date;
                    var daily = await _usageRepo.CountSinceAsync(tenantId, "api_notification_request", todayStart, policy.Channel);
                    if (daily >= policy.MaxDailyUsage.Value)
                        return new EnforcementDecision { Allowed = false, Reason = $"Daily quota exceeded: {daily}/{policy.MaxDailyUsage} requests today", Code = "DAILY_QUOTA_EXCEEDED" };
                }
                if (policy.MaxMonthlyUsage.HasValue)
                {
                    var monthStart = new DateTime(now.Year, now.Month, 1, 0, 0, 0, DateTimeKind.Utc);
                    var monthly = await _usageRepo.CountSinceAsync(tenantId, "api_notification_request", monthStart, policy.Channel);
                    if (monthly >= policy.MaxMonthlyUsage.Value)
                        return new EnforcementDecision { Allowed = false, Reason = $"Monthly quota exceeded: {monthly}/{policy.MaxMonthlyUsage} requests this month", Code = "MONTHLY_QUOTA_EXCEEDED" };
                }
            }
            return new EnforcementDecision { Allowed = true };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "UsageEvaluation: enforcement check failed, allowing request");
            return new EnforcementDecision { Allowed = true };
        }
    }

    public async Task<EnforcementDecision> CheckAttemptAllowedAsync(Guid tenantId, string channel)
    {
        try
        {
            var policies = await _rateLimitRepo.FindActivePoliciesAsync(tenantId, channel);
            if (policies.Count == 0) return new EnforcementDecision { Allowed = true };

            foreach (var policy in policies)
            {
                if (policy.MaxAttemptsPerMinute.HasValue)
                {
                    var units = channel == "email" ? new[] { "email_attempt" } : new[] { "sms_attempt" };
                    var recent = await _usageRepo.CountSinceMultipleAsync(tenantId, units, DateTime.UtcNow.AddMinutes(-1), policy.Channel);
                    if (recent >= policy.MaxAttemptsPerMinute.Value)
                        return new EnforcementDecision { Allowed = false, Reason = $"Attempt rate limit exceeded: {recent}/{policy.MaxAttemptsPerMinute} attempts in last minute", Code = "RATE_LIMIT_EXCEEDED" };
                }
            }
            return new EnforcementDecision { Allowed = true };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "UsageEvaluation: attempt check failed, allowing");
            return new EnforcementDecision { Allowed = true };
        }
    }
}
